### DISCUSSION TITLE
docs: missing clear instructions/example on how to set design tokens application wide

### DISCUSSION THREAD

**Discussion by luissalgadofreire** [microsoft/fluentui-blazor#2075] [Q&A] [answered]
First of all, thank you for yet another great addition to the Blazor ecosystem. I am looking forward to use Fluent UI Blazor in my projects.

I looked at the following sources for clear instructions on how to set design tokens at application level, and also went through design-tokens-related issues in this github repo:

- https://github.com/microsoft/fluentui-blazor
- https://www.fluentui-blazor.net/DesignTokens
- https://github.com/microsoft/fluentui-blazor/blob/dev/src/Extensions/DesignToken.Generator/DesignTokenConstants.cs
- https://learn.microsoft.com/en-us/fluent-ui/web-components/design-system/design-tokens
- https://learn.microsoft.com/en-us/fluent-ui/web-components/getting-started/styling
- https://microsoft.github.io/fluentui-design-tokens/

It is still not clear for me. I like the concept of design tokens, which  - to my knowledge - are implemented as CSS Variables in the underlying Fluent UI Web Components solution. Being able to set all the design tokens gives us very considerable control of the application's style.

My particular question below.

**Is it viable and recommended to simply:**

1) add design tokens as CSS variables to the application's  `wwwroot/css/styles.css` like:
  ```css
  :root {
      --foreground-color: #000000;
      --background-color: #FFFFFF;
      --accent-fill-rest: #0078D4;
      --accent-fill-hover: #106EBE;
      --accent-fill-active: #005A9E;
      --accent-fill-focus: #005A9E;
      --control-corner-radius: 50;
  }
  ```
2) add a reference to that css file in `index.html` like in:
  ```html
  html
  Copy code
  <!DOCTYPE html>
  <html lang="en">
  <head>
      <meta charset="utf-8" />
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <title>My Blazor App</title>
      <link href="_content/MyApp/css/styles.css" rel="stylesheet" />
      <link href="css/app.css" rel="stylesheet" />
  </head>
  <body>
      <app>Loading...</app>
      <script src="_framework/blazor.webassembly.js"></script>
  </body>
  </html>
  ```

I tried it and in my experiment it worked but required the addition of `!important` after each design token variable, because Fluent UI Blazor automatically injects these in the consuming app; and apparently the only way to override this at application-level is to give our CSS variables greater priority via this not so best-practice workaround.

Is there any thing in how Fluent UI Blazor works that does not advise setting the CSS Variables as above?

If the above is not recommended, how exactly and where in a Blazor Wasm app, for example, should we set the values of Design Tokens for site-wide application? All the examples in the docs seem to  focus on changes at component level. But how can we set them at application level in a direct straight forward way?

**Comment by vnbaaij** [comment]
Not all design tokens are 'just' CSS variables. The `accent-fill-...` variables, for example, are in the script created by what is called a recipe. They are calculated based on a base color you supply to the token and take things like contrast ratio into account. If you specify those manualyy, you'll need to take care of that yourself. You also indeed need to use the `!important` addition to overrule the script generated values.

The first code example at https://www.fluentui-blazor.net/DesignTokens#using-design-tokens-from-code explicitly mentions how you can use a token 'globally': 
```
// If you would like to change the BaseLayerLuminance  value for the whole site, you can use the WithDefault method
await BaseLayerLuminance.WithDefault((float)0.15);
```
To answer your questions:
1) No, it is not advised to do this. You set the tokens and then needed CSS variables will be created for you. If you do it yourself, you will run into things like needing the `!important`

You can look at the code for the demo site on how we use the tokens to set application wide tokens as well. See also https://github.com/microsoft/fluentui-blazor/discussions/2070
Moving this as it is not an issue but more a discussion

**Reply by luissalgadofreire** [reply]
That was an incredibly fast response. Thank you for the prompt reply and sorry for wrongly opening an issue instead of a discussion.

Understood regarding the use of CSS variables. 

My main doubt regarding:

```
// If you would like to change the BaseLayerLuminance  value for the whole site, you can use the WithDefault method
await BaseLayerLuminance.WithDefault((float)0.15);
```

Is where do we put this code to be applied application wide? The examples jump straight into those lines of code, but for those starting with Blazor and Fluent UI Blazor, they lack the context of where such code should go.

As for the code of the demo site, do you mean [https://github.com/microsoft/fluentui-blazor/tree/dev/examples/Demo](url)? Where should I look? Demo/Shared?

**Reply by luissalgadofreire** [reply]
I also saw #1044 where the same question was asked: where to place `await BaseLayerLuminance.WithDefault((float)0.15);`. That question was not exactly answered. The suggestion led to the use of the following in **App.razor**:

```
<FluentDesignSystemProvider AccentBaseColor="#dedbef" BaseLayerLuminance="0.15f">
    <Router AppAssembly="@typeof(App).Assembly">
      ...
    </Router>
</FluentDesignSystemProvider>
```

Does this expose all design tokens? If not, where should we place:

```
// If you would like to change the BaseLayerLuminance  value for the whole site, you can use the WithDefault method
await BaseLayerLuminance.WithDefault((float)0.15);
```

I understand you are working short on resources, so I appreciate how fast you replied. I also see things are moving fast. Thanks for the effort.

**Reply by luissalgadofreire** [reply]
Looking at the specification [here](https://github.com/microsoft/fluentui/blob/cb984ca549773f498b4aee348bed1ee8ce416aa4/packages/web-components/src/design-system-provider/design-system-provider.vscode.definition.json), it would seem like FluentDesignSystemProvider exposes all design tokens.

**Reply by luissalgadofreire** [reply]
I will try FluentDesignSystemProvider but it would be useful to understand where to place `WithDefault()` for global styling, since it is suggested in the docs.

**Reply by vnbaaij** [reply]
No, don't use the FluentDesignSystemProvider. It is a legacy component and doesn't expose all tokens.

Yes, the demo/shared folder contains the site source. 

As the tokens use JS Interop, they need to be called from an OnAfterRenderAsync method. Where you do that depends on things like do you also use theme switching. You can use the demo source as a reference. Theme switching complicates things a bit because you do not want a flash of unstyled content. See also https://www.fluentui-blazor.net/DesignTheme

**Reply by luissalgadofreire** [reply]
Thanks.

My understanding of DesignTheme is that it focuses on Color themes. While that should be useful, I'm looking for something that can set any token.

In #427, the following code is shown, which does as you suggest, on OnAfterRenderAsync:

```
@code {
  [Inject] 
  public AccentBaseColor AccentBaseColor { get; set; } = default!;
  
  protected override Task OnAfterRenderAsync(bool firstRender)
  {
    if (firstRender)
    {
      AccentBaseColor.WithDefault("#FF0000".ToSwatch());
    }
  }
}
```

However, it is mentioned that that code had no effect  on its own: "WithDefault does just that, setting a default value for a token. To apply a token, you need to select a Dom element where you want to set the token for."

I'm not sure I follow what would that entail, with respect to applying a token application wide. Do we still need that extra step, or is it enough to just set the default value?

**Reply by luissalgadofreire** [reply]
I can't seem to find examples of application wise use of `WithDefault` in the demo site code. Application-wide, I only see usage of FluentDesignTheme in App.razor.

```
<FluentDesignTheme StorageName="theme" />

<Router AppAssembly="@typeof(App).Assembly">
    <Found Context="routeData">
        <RouteView RouteData="@routeData" DefaultLayout="@typeof(DemoMainLayout)" />
    </Found>
    <NotFound>
        <PageTitle>Not found</PageTitle>
        <LayoutView Layout="@typeof(DemoMainLayout)">
            <p role="alert">Sorry, there's nothing at this address.</p>
        </LayoutView>
    </NotFound>
</Router> 
```

**Reply by vnbaaij** [reply]
>However, it is mentioned that that code had no effect on its own: "WithDefault does just that, setting a default value for a token. To apply a token, you need to select a Dom element where you want to set the token for."

>I'm not sure I follow what would that entail, with respect to applying a token application wide. Do we still need that extra step, or is it enough to just set the default value?

It depends on which token you are setting. In a lot of cases the effect will be reflected immediately.

**Reply by vnbaaij** [reply]
>I can't seem to find examples of application wise use of WithDefault in the demo site code. Application-wide, I only see usage of FluentDesignTheme in App.razor.

Just use the example from the design token docs I mentioned before. You'll see the value be applied once you set it. Change some of the others from that example from SetValueFor to WithDefault. Just play and experiment.

**Reply by luissalgadofreire** [reply]
It worked. I added what I did below for the sake of completion.

Thanks for your time.

**Comment by luissalgadofreire** [comment, marked as answer]
For the sake of completion.

I am using a Razor Class Library (RCL) to share UI components between platforms (Blazor WebAssembly for Web, Blazor Hybrid for Native). The objective is for this RCL to serve as an internal UI Kit, in a way that abstracts consuming cross-platform apps from knowing other libraries but the RCL. 

I was unsure of how to set **application-wide defaults** for the Design Tokens that rule the overall style of the Fluent UI Blazor Components.

After multiple surprisingly fast clarifications from @vnbaaij, I did the following.

### In the Razor Class Library

**Step 1** (specific to my use-case): I created my own component **Button.razor**, wrapping `FluentButton`:

```razor
@namespace MyRcl.Components

<FluentButton Disabled="@Disabled" @attributes="UnmatchedValues">
    @ChildContent
</FluentButton>

@code {
    [Parameter]
    [EditorRequired]
    public RenderFragment ChildContent { get; set; }

    [Parameter]
    public bool Disabled { get; set; }

    [Parameter(CaptureUnmatchedValues = true)]
    public Dictionary<string, object> UnmatchedValues { get; set; }

    // other parameters to be exposed
}
```

**Step 2**: I created razor component **DesignTokens.razor** with the following code:

  ```razor
  @using Microsoft.FluentUI.AspNetCore.Components.DesignTokens
  @namespace MyRcl.Components

  <!-- Set the default theme using builtin FluentDesignToken machinery -->  
  <!-- This is appropriate for the color theme, not for general control of tokens -->  
  <!-- See: https://www.fluentui-blazor.net/DesignTheme -->  
  <FluentDesignTheme StorageName="theme" />
  <script src="_content/Microsoft.FluentUI.AspNetCore.Components/js/loading-theme.js" type="text/javascript"></script>
  <loading-theme storage-name="theme"></loading-theme>

  @code {
      [Inject]
      private ControlCornerRadius ControlCornerRadius { get; set; } = default!;
  
      protected override async Task OnAfterRenderAsync(bool firstRender) {
          if (firstRender) {
              // WithDefault() allows us to control any exposed design token, not just the color theme
              await ControlCornerRadius.WithDefault(50); // <-- setting the default design token value here
              // other tokens to be tested
          }
      }
  
  }
  ```

**Step 3** (specific to my use-case): To have the RCL inject Fluent UI Blazor Components in order to abstract the latter from consuming apps, I created a `ServiceCollectionExtensions` class with the following code:

```c#
using Microsoft.Extensions.DependencyInjection;
using Microsoft.FluentUI.AspNetCore.Components;

namespace MyRcl.Components;

public static class ServiceCollectionExtensions {

  public static IServiceCollection AddMyUIComponents(this IServiceCollection services) {
    services.AddFluentUIComponents();

    return services;
  }

}
```

### In the consuming Blazor WebAssembly app, as an example:

**Step 4**: I added the following line in the `<head>` section of the **index.html**:

```c#
<link rel="stylesheet" href="_content/Microsoft.FluentUI.AspNetCore.Components/css/reboot.css"/>
```

NOTE: one can abstract this by having the RCL directly supply `reboot.css` or equivalent. Just add it to the RCL's `wwwroot/css` folder and then, instead of pointing to the above path, point to your RCL's file.

**Step 5** (specific to my use-case): I added the following line in **Program.cs** to inject the underlying components:

```c#
builder.Services.AddMyUIComponents();
// In normal circumstances, one will simply add `builder.Services.AddFluentUIComponents();`
```


**Step 6**:  Finally, in **App.razor**, I added the `DesignTokens` component at the top so that it would set my design token default values and color theme in the consuming app:

```razor
<DesignTokens /> @* <-- add this *@

<MyApp Title="My UI Kit"
                 Assemblies="[typeof(App).Assembly]"
                 DefaultLayout="typeof(DefaultLayout)"
</MyApp>
```

NOTE: this will fail if step 5 is missed.

### Conclusions

1) The button's border radius changed so `ControlCornerRadius.WithDefault(50);` worked.
2) The button's color style changed depending on the browser's light/dark mode, so `FluentDesignTheme` worked.
2) This solution allowed me to combine a more generic approach via `<DesignToken>.WithDefault()` with the builtin `FluentDesignTheme`.
3) The Fluent UI Blazor team is incredibly fast responding to our queries.
